Make generate_trajectories more flexible

ejnnr opened the pull request:
`generate_trajectories` currently needs an SB3 policy or algorithm. With this PR, it also accepts a `Callable` that generates actions from observations, or `policy=None` (in which case random actions are used). I don't have any plans to use this inside `imitation` itself, only in the reward preprocessing project, so if you don't think it belongs here, feel free to close this and I'll just implement it there. But #271 already wanted support for `Callable`s so maybe others will use that as well in the future, which is why I created the PR here.

AdamGleave reviewed (approved):
Thanks for the PR. This feature seems worth upstreaming into imitation even if we don't need it there yet -- it's a fairly natural thing to want to do.

I've refactored the policy to callable conversion into its own function -- I don't feel strongly about this (it's not used anywhere else), so feel free to revert that, but it seemed a bit easier to reason about and has a nice and simple type signature.

I also expanded the types of `generate_transitions` and `rollout_and_save` to allow for callables/None, too, since they just use `generate_trajectories` under the hood.

Please do check my changes before merging!

ejnnr commented:
Thanks for the review, and I like your changes :) Your reordering of the `elif` statements introduced a bug though -- luckily the test suite caught that because I wouldn't have (my original ordering wasn't deliberate). I've fixed it and added a comment (unfortunately also had to add back in the `pytype` ignore comment). I've also added some smoke tests for random and `Callable` policies.

I've requested another review in case you spot any other similar bugs, and so you can take a look at the new tests.

AdamGleave reviewed (approved):
> I've requested another review in case you spot any other similar bugs, and so you can take a look at the new tests.

LGTM.

> Your reordering of the `elif` statements introduced a bug though -- luckily the test suite caught that because I wouldn't have (my original ordering wasn't deliberate).

Ugh, sorry, should have checked the tests passed before committing. Thanks for fixing my mistake!

> Hm, not sure how to best deal with this line in the test suite that's never executed:

I just added a `# pragma: no cover` line. We demand 100% test coverage so we don't accidentally write test code that doesn't get executed. But things like this (input validation for a helper method in a test) are fine to not execute, so it's fine to override it.
